src/coordination/coordinator_instance.cpp

```diff
@@ -383,61 +379,145 @@ auto CoordinatorInstance::ForceResetCluster_() -> ForceResetClusterStateStatus {
 
-    repl_instances_.emplace_back(std::move(client), &CoordinatorInstance::ReplicaSuccessCallback,
-                                 &CoordinatorInstance::ReplicaFailCallback);
+    repl_instances_.emplace_back(std::move(client));
   });
 
   auto instances_mapped_to_resp = repl_instances_ | ranges::views::transform([](auto &instance) {
                                     return std::pair{instance.InstanceName(), instance.SendFrequentHeartbeat()};
                                   }) |
                                   ranges::to<std::unordered_map<std::string, bool>>();
 
+  auto const maybe_main_instance =
+      std::ranges::find_if(repl_instances_, [this, &instances_mapped_to_resp](auto const &instance) {
+        return instances_mapped_to_resp[instance.InstanceName()] && raft_state_->IsCurrentMain(instance.InstanceName());
+      });
+
+  auto const have_alive_main_instance = maybe_main_instance != repl_instances_.end();
+  if (have_alive_main_instance) {
+    spdlog::trace("Alive MAIN instance found {}.", maybe_main_instance->InstanceName());
+    // if we have alive MAIN instance alive we expect cluster was in correct state already, we can start frequent checks
+    // and set callbacks
+
+    auto alive_instances = repl_instances_ | ranges::views::filter([&instances_mapped_to_resp](auto const &instance) {
+                             return instances_mapped_to_resp[instance.InstanceName()];
+                           });
+
+    auto alive_replica_instances = alive_instances | ranges::views::filter([this](auto const &instance) {
+                                     return raft_state_->HasReplicaState(instance.InstanceName());
+                                   });
+
+    auto alive_non_replica_instances = alive_instances | ranges::views::filter([this](auto const &instance) {
+                                         return !raft_state_->HasReplicaState(instance.InstanceName());
+                                       });
+
+    // our set of instances currently consists of
+    // 1. Alive MAIN
+    // 2. Alive REPLICAS
+    // 3. Alive non-replicas (instances which need to be demoted)
+    // 4. Dead REPLICA(s)
+    // 5. Dead MAIN(s)
+
+    // For instances 3, 4 and 5 we need to set demote callbacks
+
+    std::ranges::for_each(alive_replica_instances, [](auto &instance) {
+      instance.SetCallbacks(&CoordinatorInstance::ReplicaSuccessCallback, &CoordinatorInstance::ReplicaFailCallback);
+    });
+
+    std::ranges::for_each(alive_non_replica_instances, [](auto &instance) {
+      instance.SetCallbacks(&CoordinatorInstance::DemoteSuccessCallback, &CoordinatorInstance::DemoteFailCallback);
+    });
+
+    auto dead_instances = repl_instances_ | ranges::views::filter([&instances_mapped_to_resp](auto const &instance) {
+                            return !instances_mapped_to_resp[instance.InstanceName()];
+                          });
+
+    std::ranges::for_each(dead_instances, [this](auto &instance) {
```

Comment:
```suggestion
    std::ranges::for_each(dead_instances, [](auto &&instance) {
```

Reply:
From the moment since we find which instances are dead to the moment of setting correct callbacks, instance could die. Could we get into problems because of it? Maybe it would be better that we set to main instance promote callback and to replica instances demote callback, more safe I think

Reply:
resolved